File: src/test/java/com/github/nsc/de/compiler/parser/ParserTests.java
package com.github.nsc.de.compiler.parser;

import static org.junit.jupiter.api.Assertions.*;
import static com.github.nsc.de.compiler.parser.ParserTestUtil.*;
import static com.github.nsc.de.compiler.TestUtil.*;

import com.github.nsc.de.compiler.parser.node.*;
import com.github.nsc.de.compiler.parser.node.expression.NumberNode;
import com.github.nsc.de.compiler.parser.node.logical.LogicalSmallerNode;
import com.github.nsc.de.compiler.parser.node.logical.LogicalTrueNode;
import com.github.nsc.de.compiler.parser.node.loops.DoWhileNode;
import com.github.nsc.de.compiler.parser.node.loops.ForNode;
import com.github.nsc.de.compiler.parser.node.loops.WhileNode;
import com.github.nsc.de.compiler.parser.node.variables.*;
import org.junit.jupiter.api.Test;

public class ParserTests {

    @Test
    public void testMultiStatement() {

        Tree node = parse("<MultiStatementTest>", "var i; i");

        assertEquals(2, node.getChildren().length);
        assertType(VariableDeclarationNode.class, node.getChildren()[0]);
        assertType(VariableUsageNode.class, node.getChildren()[1]);

        node = parse("<MultiStatementTest>", "var i\ni");

        assertEquals(2, node.getChildren().length);
        assertType(VariableDeclarationNode.class, node.getChildren()[0]);
        assertType(VariableUsageNode.class, node.getChildren()[1]);

    }

    @Test
    public void testWhile() {
        WhileNode node = parseSingle("<WhileTest>", "while (true) { i }", WhileNode.class);
        assertNotNull(node.getCondition());
        assertNotNull(node.getBody());

        assertType(LogicalTrueNode.class, node.getCondition());
        assertType(Tree.class, node.getBody());


        node = parseSingle("<WhileTest>", "while (true) i;", WhileNode.class);
        assertNotNull(node.getCondition());
        assertNotNull(node.getBody());

        assertType(LogicalTrueNode.class, node.getCondition());
        assertType(Tree.class, node.getBody());
    }

    @Test
    public void testDoWhile() {
        DoWhileNode node = parseSingle("<DoWhileTest>", "do { i } while (true);", DoWhileNode.class);
        assertNotNull(node.getCondition());
        assertNotNull(node.getBody());

        assertType(LogicalTrueNode.class, node.getCondition());
        assertType(Tree.class, node.getBody());


        node = parseSingle("<WhileTest>", "do i; while (true);", DoWhileNode.class);
        assertNotNull(node.getCondition());
        assertNotNull(node.getBody());

        assertType(LogicalTrueNode.class, node.getCondition());
        assertType(Tree.class, node.getBody());
    }

    @Test
    public void testFor() {
        ForNode node = parseSingle("<ForTest>", "for(var i = 0; i < 100; i++) { i; }", ForNode.class);
        assertNotNull(node.getDeclaration());
        assertNotNull(node.getCondition());
        assertNotNull(node.getRound());
        assertNotNull(node.getBody());

        assertType(VariableDeclarationNode.class, node.getDeclaration());
        assertType(LogicalSmallerNode.class, node.getCondition());
        assertType(VariableIncreaseNode.class, node.getRound());
        assertType(Tree.class, node.getBody());


        node = parseSingle("<ForTest>", "for(var i = 0; i < 100; i++) i;", ForNode.class);
        assertNotNull(node.getDeclaration());
        assertNotNull(node.getCondition());
        assertNotNull(node.getRound());
        assertNotNull(node.getBody());

        assertType(VariableDeclarationNode.class, node.getDeclaration());
        assertType(LogicalSmallerNode.class, node.getCondition());
        assertType(VariableIncreaseNode.class, node.getRound());
        assertType(Tree.class, node.getBody());
    }

    @Test
    public void testIf() {
        IfNode node = parseSingle("<IfTest>", "if (true) { i }", IfNode.class);
        assertNotNull(node.getCondition());
        assertNotNull(node.getBody());

        assertType(LogicalTrueNode.class, node.getCondition());
        assertType(Tree.class, node.getBody());


        node = parseSingle("<IfTest>", "if (true) i;", IfNode.class);
        assertNotNull(node.getCondition());
        assertNotNull(node.getBody());
        assertNull(node.getElseBody());
        assertType(LogicalTrueNode.class, node.getCondition());
        assertType(Tree.class, node.getBody());


        node = parseSingle("<IfTest>", "if (true) i; else i;", IfNode.class);
        assertNotNull(node.getCondition());
        assertNotNull(node.getBody());
        assertNotNull(node.getElseBody());
        assertType(LogicalTrueNode.class, node.getCondition());
        assertType(Tree.class, node.getBody());
        assertType(Tree.class, node.getElseBody());


        node = parseSingle("<IfTest>", "if (true) i; else if (true) i; else i;", IfNode.class);
        assertNotNull(node.getCondition());
        assertNotNull(node.getBody());
        assertNotNull(node.getElseBody());
        assertType(LogicalTrueNode.class, node.getCondition());
        assertType(Tree.class, node.getBody());
        assertType(Tree.class, node.getElseBody());

        Tree elseBody = node.getElseBody();
        assertSame(1, elseBody.getChildren().length);
        assertType(IfNode.class, elseBody.getChildren()[0]);

        IfNode if2 = (IfNode) elseBody.getChildren()[0];
        assertType(Tree.class, if2.getBody());
        assertType(Tree.class, if2.getElseBody());


    }

}

File: src/test/kotlin/com/github/nsc/de/shake/parser/ParserTests.kt
package com.github.nsc.de.shake.parser

import com.github.nsc.de.shake.assertType
import com.github.nsc.de.shake.parser.node.IfNode
import com.github.nsc.de.shake.parser.node.ImportNode
import com.github.nsc.de.shake.parser.node.Tree
import com.github.nsc.de.shake.parser.node.logical.LogicalSmallerNode
import com.github.nsc.de.shake.parser.node.logical.LogicalTrueNode
import com.github.nsc.de.shake.parser.node.loops.DoWhileNode
import com.github.nsc.de.shake.parser.node.loops.ForNode
import com.github.nsc.de.shake.parser.node.loops.WhileNode
import com.github.nsc.de.shake.parser.node.variables.VariableDeclarationNode
import com.github.nsc.de.shake.parser.node.variables.VariableIncreaseNode
import com.github.nsc.de.shake.parser.node.variables.VariableUsageNode
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class ParserTests {
    @Test
    fun testImport() {
        var node = ParserTestUtil.parse("<ImportTest>", "import com;")
        Assertions.assertEquals(1, node.children.size)
        assertType(ImportNode::class.java, node.children[0])
        Assertions.assertArrayEquals(arrayOf("com"), (node.children[0] as ImportNode).import)
        node = ParserTestUtil.parse("<ImportTest>", "import com.github.nsc.de.shake;")
        Assertions.assertEquals(1, node.children.size)
        assertType(ImportNode::class.java, node.children[0])
        Assertions.assertArrayEquals(
            arrayOf("com", "github", "nsc", "de", "shake"),
            (node.children[0] as ImportNode).import
        )
        node = ParserTestUtil.parse("<ImportTest>", "import com.github.nsc.de.shake.*;")
        Assertions.assertEquals(1, node.children.size)
        assertType(ImportNode::class.java, node.children[0])
        Assertions.assertArrayEquals(
            arrayOf("com", "github", "nsc", "de", "shake", "*"),
            (node.children[0] as ImportNode).import
        )
    }

    @Test
    fun testMultiStatement() {
        var node = ParserTestUtil.parse("<MultiStatementTest>", "var i; i")
        Assertions.assertEquals(2, node.children.size)
        assertType(VariableDeclarationNode::class.java, node.children[0])
        assertType(VariableUsageNode::class.java, node.children[1])
        node = ParserTestUtil.parse("<MultiStatementTest>", "var i\ni")
        Assertions.assertEquals(2, node.children.size)
        assertType(VariableDeclarationNode::class.java, node.children[0])
        assertType(VariableUsageNode::class.java, node.children[1])
    }

    @Test
    fun testWhile() {
        var node = ParserTestUtil.parseSingle("<WhileTest>", "while (true) { i }", WhileNode::class.java)
        Assertions.assertNotNull(node.condition)
        Assertions.assertNotNull(node.body)
        assertType(LogicalTrueNode::class.java, node.condition)
        assertType(Tree::class.java, node.body)
        node = ParserTestUtil.parseSingle("<WhileTest>", "while (true) i;", WhileNode::class.java)
        Assertions.assertNotNull(node.condition)
        Assertions.assertNotNull(node.body)
        assertType(LogicalTrueNode::class.java, node.condition)
        assertType(Tree::class.java, node.body)
    }

    @Test
    fun testDoWhile() {
        var node = ParserTestUtil.parseSingle("<DoWhileTest>", "do { i } while (true);", DoWhileNode::class.java)
        Assertions.assertNotNull(node.condition)
        Assertions.assertNotNull(node.body)
        assertType(LogicalTrueNode::class.java, node.condition)
        assertType(Tree::class.java, node.body)
        node = ParserTestUtil.parseSingle("<WhileTest>", "do i; while (true);", DoWhileNode::class.java)
        Assertions.assertNotNull(node.condition)
        Assertions.assertNotNull(node.body)
        assertType(LogicalTrueNode::class.java, node.condition)
        assertType(Tree::class.java, node.body)
    }

    @Test
    fun testFor() {
        var node = ParserTestUtil.parseSingle("<ForTest>", "for(var i = 0; i < 100; i++) { i; }", ForNode::class.java)
        Assertions.assertNotNull(node.declaration)
        Assertions.assertNotNull(node.condition)
        Assertions.assertNotNull(node.round)
        Assertions.assertNotNull(node.body)
        assertType(VariableDeclarationNode::class.java, node.declaration)
        assertType(LogicalSmallerNode::class.java, node.condition)
        assertType(VariableIncreaseNode::class.java, node.round)
        assertType(Tree::class.java, node.body)
        node = ParserTestUtil.parseSingle("<ForTest>", "for(var i = 0; i < 100; i++) i;", ForNode::class.java)
        Assertions.assertNotNull(node.declaration)
        Assertions.assertNotNull(node.condition)
        Assertions.assertNotNull(node.round)
        Assertions.assertNotNull(node.body)
        assertType(VariableDeclarationNode::class.java, node.declaration)
        assertType(LogicalSmallerNode::class.java, node.condition)
        assertType(VariableIncreaseNode::class.java, node.round)
        assertType(Tree::class.java, node.body)
    }

    @Test
    fun testIf() {
        var node = ParserTestUtil.parseSingle("<IfTest>", "if (true) { i }", IfNode::class.java)
        Assertions.assertNotNull(node.condition)
        Assertions.assertNotNull(node.body)
        assertType(LogicalTrueNode::class.java, node.condition)
        assertType(Tree::class.java, node.body)
        node = ParserTestUtil.parseSingle("<IfTest>", "if (true) i;", IfNode::class.java)
        Assertions.assertNotNull(node.condition)
        Assertions.assertNotNull(node.body)
        Assertions.assertNull(node.elseBody)
        assertType(LogicalTrueNode::class.java, node.condition)
        assertType(Tree::class.java, node.body)
        node = ParserTestUtil.parseSingle("<IfTest>", "if (true) i; else i;", IfNode::class.java)
        Assertions.assertNotNull(node.condition)
        Assertions.assertNotNull(node.body)
        Assertions.assertNotNull(node.elseBody)
        assertType(LogicalTrueNode::class.java, node.condition)
        assertType(Tree::class.java, node.body)
        assertType(Tree::class.java, node.elseBody)
        node = ParserTestUtil.parseSingle("<IfTest>", "if (true) i; else if (true) i; else i;", IfNode::class.java)
        Assertions.assertNotNull(node.condition)
        Assertions.assertNotNull(node.body)
        Assertions.assertNotNull(node.elseBody)
        assertType(LogicalTrueNode::class.java, node.condition)
        assertType(Tree::class.java, node.body)
        assertType(Tree::class.java, node.elseBody)
        val elseBody = node.elseBody
        Assertions.assertSame(1, elseBody.children.size)
        assertType(IfNode::class.java, elseBody.children[0])
        val if2 = elseBody.children[0] as IfNode
        assertType(Tree::class.java, if2.body)
        assertType(Tree::class.java, if2.elseBody)
    }
}